File: backend/app/services/vector_store.py

import math

# Global In-Memory Store
# Structure: List of dicts: {"id": str, "text": str, "embedding": list[float], "doc_id": str}
store = []

def add_to_store(chunk):
    """
    chunk: dict with keys 'id', 'text', 'embedding'
    """
    store.append(chunk)

def get_store():
    return store

def clear_store():
    global store
    store = []

def cosine_similarity(a, b):
    dot = sum(x * y for x, y in zip(a, b))
    norm_a = math.sqrt(sum(x * x for x in a))
    norm_b = math.sqrt(sum(x * x for x in b))
    if norm_a == 0 or norm_b == 0:
        return 0
    return dot / (norm_a * norm_b)

def search_similar(query_embedding, top_k=3):
    """
    Returns top_k chunks sorted by cosine similarity
    """
    scores = []
    for chunk in store:
        score = cosine_similarity(query_embedding, chunk['embedding'])
        scores.append({
            "text": chunk['text'],
            "score": score,
            "id": chunk.get('id')
        })
    
    # Sort descending
    scores.sort(key=lambda x: x['score'], reverse=True)
    return scores[:top_k]

File: backend/_old_ts/src/services/vectorStore.ts
export type Chunk = {
  id:string,
  text:string,
  embedding:number[]
}

let store:Chunk[] = []

export function addToStore(chunk:Chunk){
  store.push(chunk)
}

export function getStore(){
  return store
}

export function cosineSim(a:number[],b:number[]){
  let dot=0, normA=0, normB=0
  for(let i=0;i<a.length;i++){
    dot+=a[i]*b[i]
    normA+=a[i]*a[i]
    normB+=b[i]*b[i]
  }
  if(normA === 0 || normB === 0) return 0
  return dot/(Math.sqrt(normA)*Math.sqrt(normB))
}

export function searchSimilar(queryEmbedding:number[]){
  const scored = store.map(c=>({
    text:c.text,
    score:cosineSim(queryEmbedding,c.embedding)
  }))

  return scored.sort((a,b)=>b.score-a.score).slice(0,3)
}
